gtime/plotting/preprocessing.py

```diff
@@ -0,0 +1,199 @@
+            week = t.start_time.weekofyear
+        else:
+            year = t.start_time.year
+            week = t.start_time.weekofyear
+        return "_".join([str(year), str(week)])
+
+    if freq is None:
+        freq = df.index.freqstr
+    df = df.resample(freq).agg(agg)
+    col_name = df.columns[0]
+
+    if cycle in ["year", "quarter", "month", "week"]:
+        if cycle == "year":
+            df["_Series"] = df.index.start_time.year
+            if freq == "D":
+                df["_Season"] = df.index.dayofyear
+            elif freq in ["W", "W-SUN"]:
+                df["_Season"] = df.start_time.weekofyear
+            elif freq == "M":
+                df["_Season"] = df.index.start_time.month
+            elif freq in ["Q", "Q-DEC"]:
+                df["_Season"] = df.index.start_time.quarter
+            else:
+                df["_Season"] = (
+                    df[col_name]
+                    .groupby(pd.Grouper(freq="Y", convention="s"))
+                    .cumcount()
+                    + 1
+                )
+
+        elif cycle == "quarter":
+            df["_Series"] = list(
+                map(
+                    lambda x: "_".join(
+                        [str(x.start_time.year), str(x.start_time.quarter)]
+                    ),
+                    df.index,
+                )
+            )
+            df["_Season"] = (
+                df[col_name].groupby(pd.Grouper(freq="Q", convention="s")).cumcount()
+                + 1
+            )
+
+        elif cycle == "month":
+            df["_Series"] = list(
+                map(
+                    lambda x: "_".join(
+                        [str(x.start_time.year), str(x.start_time.month)]
+                    ),
+                    df.index,
+                )
+            )
+            if freq == "D":
+                df["_Season"] = df.index.day
+            else:
+                df["_Season"] = (
+                    df[col_name]
+                    .groupby(pd.Grouper(freq="M", convention="s"))
+                    .cumcount()
+                    + 1
+                )
+
+        elif cycle == "week":
+            df["_Series"] = list(map(_week_of_year, df.index))
+            if freq == "D":
+                df["_Season"] = df.index.dayofweek
+            else:
+                df["_Season"] = (
+                    df[col_name]
+                    .groupby(pd.Grouper(freq="W", convention="s"))
+                    .cumcount()
+                    + 1
+                )
+        else:
+            raise ValueError("Incorrect cycle period name")
+    else:
+        df["_Series"] = (
+            df[col_name].groupby(pd.Grouper(freq=cycle, convention="s")).ngroup()
+        )
+        df["_Season"] = (
+            df[col_name].groupby(pd.Grouper(freq=cycle, convention="s")).cumcount() + 1
+        )
+
+    return df.set_index(["_Series", "_Season"]).unstack(level=0)
+
+
+def acf(x, max_lags=None):
```

Comment:
In general, try to split functions in smaller and clearer functions. For example, this function can be rewritte in:
    
    x = _normalize(x)
    all_lags_acf = _autocorrelate_for_all_lags(x)
    acf = _select_first_n_lags(all_lags_acf, max_lags)

Or something similar, so that you can immediately get what each step is doing.
Try to apply this principle for each function, it is also much easier to test.